### ISSUE TITLE
help making an autoencoder

### ISSUE THREAD

**Issue by stablum** [Lasagne/Lasagne#461]
Hi,

I have tweaked around with Lasagne and I found it an excellent library. My experiments revolve around building an autoencoder for the MNIST dataset. Unfortunately I didn't manage to train it properly, maybe you can spot any mistakes I made?

Here is the structure I chose:

 * input layer (28x28)
 * 2D convolutional layer, filter size 7x7
 * Max Pooling layer, size 3x3, stride 2x2
 * Dense (fully connected) flattening layer, 10 units (this is the bottleneck)
 * Dense (fully connected) layer, 121 units
 * Reshaping layer to 11x11
 * 2D convolutional layer, filter size 3x3
 * 2D Upscaling layer factor 2
 * 2D convolutional layer, filter size 3x3
 * 2D Upscaling layer factor 2
 * 2D convolutional layer, filter size 5x5
 * Feature max pooling (from 31x28x28 to 28x28)

All the 2D convolutional layers have the biases untied, sigmoid activations and 31 filters.

All the fully connected layers have sigmoid activations.

The loss function used is squared error, the updating function is adagrad. The length of the chunk for the learning is 100 samples, multiplied for 1000 epochs.

here is the original post I made on *stats.stackexchange.com* (containing the full source code of the autoencoder): http://stats.stackexchange.com/questions/176881/cannot-make-this-autoencoder-network-function-properly-with-convolutional-and-m

Thank you for any hint on what might be wrong!

**Comment by f0k** [comment]
Hello, please re-post this on our mailing list: https://groups.google.com/forum/#!forum/lasagne-users
The github issue tracker is reserved for bug reports, feature discussions and implementations.

To still answer something: I'd suggest to start by reproducing the results of some paper (e.g., http://people.idsia.ch/~ciresan/data/icann2011.pdf) using their architecture. Your bottleneck seems to come a bit too sudden to me.

**Comment by stablum** [comment]
Thank you the hints. I actually managed to make the autoencoder work with only two bottleneck nodes. I have posted the implementation here: http://stats.stackexchange.com/questions/176881/cannot-make-this-autoencoder-network-function-properly-with-convolutional-and-m

Would you be interested into adding it to the examples? I might rewrite the code in a nicer way if you wish to add it.

**Comment by f0k** [comment]
> I actually managed to make the autoencoder work with only two bottleneck nodes.

Fun how it likes to turn almost everything into nines :)

> Would you be interested into adding it to the examples?

We'd definitely like a convolutional auto-encoder in [Lasagne/Recipes](https://github.com/Lasagne/Recipes), but I think a conventional symmetric architecture would be suited better. Or maybe a script similar to the mnist example in Lasagne that allows to try a couple of different architectures (one written like yours, but symmetrical, one with tied weights using `InverseLayer`, and maybe one unconventional). It should probably also demonstrate how to decode from a given code, as that has been asked on the mailing list. So... it's a bit more work than cleaning up your code... if you're interested in giving it a go, give it a go :)

**Comment by stablum** [comment]
Maybe the nine encompasses characteristics of most other digits. I suspect that the nine is a fuzzy "or" of, for example, 3s, 7s, and 4s. Interestingly (and correctly) the "wrong" part of the nine, for a digit that is not nine, is less intense. Maybe the problem can be mitigated by an additional 3x3 or 5x5 convolution at the very end, which might act like an edge detector.

I wonder why a symmetrical architecture would be better. Does it lead to a reduction in reconstruction error or maybe some other theoretical aspects?

**Comment by f0k** [comment]
> I wonder why a symmetrical architecture would be better. Does it lead to a reduction in reconstruction error or maybe some other theoretical aspects?

Well, it probably doesn't have to be exactly symmetrical, but in your architecture, the decoder is a lot more expressive than the encoder. You're basically training an elaborate decoder to make sense of codes hastily thrown together. The analogy may be a bit far-fetched, but in regular compression algorithms, encoding takes a lot more cycles than decoding -- so if anything, I'd expect the decoder to have fewer layers, not the encoder.

**Comment by stablum** [comment]
That's interesting. So this means that my low dimensional representation is probably of low-quality and the decoder is trying, with some success, to make sense of them. So, by adding some more intermediate Conv/maxpool layers in the encoder, I should be able to get a better reconstruction, as well as a nicer low-dimensional representation?

**Comment by f0k** [comment]
> So, by adding some more intermediate Conv/maxpool layers in the encoder, I should be able to get a better reconstruction, as well as a nicer (semantic) low-dimensional representation?

That's what I'd expect, yes. Try more smaller convolutions instead of a single 7x7 one. And note that every pooling layer throws away spatial information that could make it harder for the decoder, so more pooling is not necessarily better.

**Comment by ToniCreswell** [comment]
https://github.com/ToniCreswell/ConvolutionalAutoEncoder

**Comment by f0k** [comment]
Thanks for sharing! For more difficult datasets than MNIST, you might want to look at #862 (possibly also for strided convolutions in the encoder).
